Introduce Scraper interface

arvenil opened the pull request:
# Introduces Scraper interface:
```go
type Scraper interface {
	Scrape(db *sql.DB, ch chan<- prometheus.Metric) error
	Name() string
	Help() string
}
```

# Changes:
- Decouple exporter from scrapes. Pass scrapes instead of configuration to exporter. This should also allow easier extending in forks with new scrapes.
  ```patch
  -func New(dsn string, collect Collect) *Exporter {
  +func New(dsn string, scrapers []Scraper) *Exporter {
  ```
- Run in parallel all scrapes (still using single db connection).
  ```go
  wg := &sync.WaitGroup{}
  defer wg.Wait()
  for _, scraper := range e.scrapers {
  	wg.Add(1)
  	go func(scraper Scraper) {
  		defer wg.Done()
  		label := "collect." + scraper.Name()
  		scrapeTime := time.Now()
  		if err := scraper.Scrape(db, ch); err != nil {
  			log.Errorln("Error scraping for "+label+":", err)
  			e.scrapeErrors.WithLabelValues(label).Inc()
  			e.error.Set(1)
  		}
  		ch <- prometheus.MustNewConstMetric(scrapeDurationDesc, prometheus.GaugeValue, time.Since(scrapeTime).Seconds(), label)
  	}(scraper)
   }
  ```
- Writing new scrape means just implementing `collector.Scraper` interface e.g. like this:
  ```go
  // ScrapeEngineTokudbStatus scrapes from `SHOW ENGINE TOKUDB STATUS`.
  type ScrapeEngineTokudbStatus struct{}
  
  // Name of the Scraper.
  func (ScrapeEngineTokudbStatus) Name() string {
  	return "engine_tokudb_status"
  }
  
  // Help returns additional information about Scraper.
  func (ScrapeEngineTokudbStatus) Help() string {
  	return "Collect from SHOW ENGINE TOKUDB STATUS"
  }
  
  // Scrape collects data.
  func (ScrapeEngineTokudbStatus) Scrape(db *sql.DB, ch chan<- prometheus.Metric) error {
  	...
  	return nil
  }
  ```
  and adding it in `main.go` to the list of available `scrapers` with information if it should be enabled by default:
  ```go
  // scrapers lists all possible collection methods and if they should be enabled by default.
  var scrapers = map[collector.Scraper]bool{
  	collector.ScrapeGlobalStatus{}:                true,
  	collector.ScrapeGlobalVariables{}:             true,
  	collector.ScrapeSlaveStatus{}:                 true,
  	collector.ScrapeProcesslist{}:                 false,
  	collector.ScrapeTableSchema{}:                 true,
  	collector.ScrapeInfoSchemaInnodbTablespaces{}: false,
  	collector.ScrapeInnodbMetrics{}:               false,
  	collector.ScrapeAutoIncrementColumns{}:        false,
  	collector.ScrapeBinlogSize{}:                  false,
  	collector.ScrapePerfTableIOWaits{}:            false,
  	collector.ScrapePerfIndexIOWaits{}:            false,
  	collector.ScrapePerfTableLockWaits{}:          false,
  	collector.ScrapePerfEventsStatements{}:        false,
  	collector.ScrapePerfEventsWaits{}:             false,
  	collector.ScrapePerfFileEvents{}:              false,
  	collector.ScrapePerfFileInstances{}:           false,
  	collector.ScrapeUserStat{}:                    false,
  	collector.ScrapeClientStat{}:                  false,
  	collector.ScrapeTableStat{}:                   false,
  	collector.ScrapeQueryResponseTime{}:           false,
  	collector.ScrapeEngineTokudbStatus{}:          false,
  	collector.ScrapeEngineInnodbStatus{}:          false,
  	collector.ScrapeHeartbeat{}:                   false,
  }
  ```
- Flags to enable/disable scrapes are generated from info provided by scraper.
- Alternative approach to #254 (https://github.com/prometheus/mysqld_exporter/pull/254#issuecomment-356876468).
- Covers partially #253 (without one db pool, db pool can be introduced in another PR).
- Covers partially #236, there is few things we can still merge from there.
- `scrapeTime` was not reset for `e.collect.InnodbMetrics`
- Fixed metric descriptors in new collector/perf_schema_replication_group_member_stats.go https://github.com/prometheus/mysqld_exporter/pull/265/files#diff-85361993ed6875151aa631eae17ceaa6L61

SuperQ commented:
Thanks!  I'm still traveling from FOSDEM, will take a closer look at this when I get home.

SuperQ commented:
Ping @juliusv, can you review this?

krasi-georgiev commented:
@arvenil unfortunately this would need a rebase and probably DCO signoff.

arvenil commented:
I can re-base the code and could you point me where to sign?

krasi-georgiev commented:
after you rebase you can just do
`git commit -s --amend` - update the last commit to add the signature

or for a new commit 
`git commit -s`

these are just to add the Signed-off-by: line to the commit message.

SuperQ commented:
Ping, let's rebase, sign, and we can get this merged.

arvenil commented:
@krasi-georgiev @SuperQ 
[x] DCO sign-off
[x] rebased
[x] `scrapers = filteredScrapers` removed, was a bug, good catch, I've fixed that in percona/mysqld_exporter for, but forgot here. Done. Thanks.
[x] Fixed metric descriptors in new  collector/perf_schema_replication_group_member_stats.go https://github.com/prometheus/mysqld_exporter/pull/265/files#diff-85361993ed6875151aa631eae17ceaa6L61 (meh, I know I should probably make separate PR for that but was super quick). Also can you guys ask people to provide tests for new scraper and functionalities? I'm not asking to require them, just ask if they could be so nice to provide some - this would help detect such bugs earlier.

SuperQ commented:
Yes, I agree, we need to increase our test coverage.  Maybe we should add a coverage bot.

krasi-georgiev commented:
not sure what changed since my last review , but if you only rebased and removed the  `scrapers = filteredScrapers` that it should be OK :wink: 

I guess in the future I will review only after a rebase and require only incremental commits to make it easy to view any changes after my initial review.
